### PR TITLE
Tile

### PR DESCRIPTION
**Pull request by tc975**
Tile and TileTest complete except for the parts that require Player and Roboticon

### CODE REVIEW

File: RoboticonColony/Assets/Code/PhysicalTile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/PhysicalTile.cs]: Is this a typo?

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: Perhaps rather than an ownerId we should have a reference to the owners object

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: if we change this to be a reference to the owner rather than an id we should make this null

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: maybe roboticons should be kept separately from resources as they arent the same catergory

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: I think this method should just return a production value for the resources as adding the resources to the player inventory isnt really the concern of a tile

File: RoboticonColony/Assets/Editor/Unit_Tests/TileTest.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Editor/Unit_Tests/TileTest.cs]: We should implement these tests before merging . We should also have some tests for when things go wrong, ie assert errors when someone tries to use AddRoboticon invalidly and so on

**Review by RobinStephenson — CHANGES_REQUESTED** [review]
Still needs things like rendering in the scene. Probably should merge the class into the mono behaviour as well.
Definitely a good start though and most problems are easy fixes

**Comment by Mikeywalsh** [comment]
Rendering is done automatically through unity, if we make tile a monobehavior then we will have to rewrite some of it, as monobehaviors do not behave like classes:

1) Monibehaviors need a corresponding game object in the scene to exist, the game object handles rendering and has other components such as transform
2) Monobehaviors exist when the game starts or if they are created at runtime, they do not have constructors, and instead use the Start() method which handles all their creation logic

If you do combine both tile classes into one, then I guess you could handle creating the roboticon Sprite over a tile if it has one on it. If roboticons have more than one Sprite for different customisations then Matthew would need to include a path to the Sprite as a member of roboticon. All sprites we load at runtime go in a folder we call 'Resources' in the unity project root. I know it's a lot to take in, if you want I can take over and do the unity specific stuff

**Comment by RobinStephenson** [comment]
Sorry I should have been clearer what i meant by rendering. I mean we need methods to change the appearance of the tile, for example to set it as the tile currently selected by the user, then the tile class should update its sprite to one with a highlight or something. And it should automatically change sprites for installed roboticons etc when one is installed or removed.
As for keeping it separate from the monobehaviour that's fine with me if you think its better this way.

**Comment by RobinStephenson** [comment]
Also did you mean to merge into Tile-Player-Robo rather than master?

**Comment by Mikeywalsh** [comment]
That would be my fault since I told him to. I thought it would be better to have you three put your classes into another branch and finish them off fully before finally pushing to master since master should always be a working version of the project.
But i've ever really done a proper group git project before so im unsure if this is actually ever done or not, sorry if it isn't!

**Comment by RobinStephenson** [comment]
No that's a good idea! I was just checking. I think we just need the tests to be finished and the other changes I mentioned unless anyone disagrees

**Comment by Mart500** [comment]
Would it not be easier to just add sprites like an overlay rather than
change the tile sprite. Whichever is easier.

**Comment by RobinStephenson** [comment]
Yeah just rendering a highlight on top of the tile makes more sense but we still need the methods to do it.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [inline RoboticonColony/Assets/Code/Tile.cs]: If owner ID is changed to a reference this needs to change also too.

private (ownerclass) _selectedOwner = new (ownerclass) 

Something like this?

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [inline RoboticonColony/Assets/Code/Tile.cs]: Better to merge these?

public int GetResource(ItemType resourceType){
         return _Resources[resourceType];
}

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mikeywalsh) [reply RoboticonColony/Assets/Code/Tile.cs]: This wouldn't even compile. _Resources is an itemtype>int dictionary, you can't, and shouldn't try to assign roboticon to it, as it makes no sense. ItemType.Roboticon is only ever used for keeping track of how many roboticons the player or market has in their inventory before they actually deploy it onto a tile.

**Review by Mikeywalsh — CHANGES_REQUESTED** [review]
You're not using ItemType like it should be, you'll need references to an actual Roboticon object in the class.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [reply RoboticonColony/Assets/Code/Tile.cs]: You wouldnt want a new instance of the owner, just pass in a reference to a player class

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mikeywalsh) [inline RoboticonColony/Assets/Code/Tile.cs]: Same here. You should be setting a reference of roboticon to be null.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [reply RoboticonColony/Assets/Code/Tile.cs]: I think both are valid. Might be useful to have both, your suggestion to allow enumeration, and the existing implementation for readability

File: RoboticonColony/ProjectSettings/ProjectVersion.txt
Comment (RobinStephenson) [inline RoboticonColony/ProjectSettings/ProjectVersion.txt]: You need to update your version of Unity if youre on your own PC

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: Shouldnt comment out code. just remove it

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: Still using python style _Name for private stuff

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: OwnerID should be changed into Owner of type AbstractPlayer

**Review by RobinStephenson — CHANGES_REQUESTED** [review]
Hard to do the review with code commented out as its hard to see whats code and whats not. remove it and ill have another look.
As well there are some other changes that need doing such as still having ownerid rather than owner and python style private vars

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: You dont need to have both a public and a private installed roboticon

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: ProduceOre and ProducePower are just the same as the respective properties

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: Player has a private set but this method which lets you set without any checks. Better to just make player public or remove this method

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: Dont see why private member has a get and set?

File: RoboticonColony/Assets/Code/Tile.cs
Comment (RobinStephenson) [inline RoboticonColony/Assets/Code/Tile.cs]: For add and remove roboticon see comment above about installedRoboticon being private/public as im not sure what its meant to be at the moment

**Review by RobinStephenson — CHANGES_REQUESTED** [review]

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [reply RoboticonColony/Assets/Code/Tile.cs]: If you set it public then this function becomes redundant

**Review by Mart500 — COMMENTED** [review]
Just some small stuff

**Review by RobinStephenson — APPROVED** [review]

**Comment by RobinStephenson** [comment]
@Mikeywalsh What do you think of the new changes? I think anything else we can iron out when we merge the other classes in

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [reply RoboticonColony/Assets/Code/Tile.cs]: Now that installedRoboticon is public this is redundant

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [inline RoboticonColony/Assets/Code/Tile.cs]: Calculation of production is already handled within the Roboticon class. Function should be removed.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [reply RoboticonColony/Assets/Code/Tile.cs]: Now Owner is public and so is the setter, This is redundant and should be removed.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [inline RoboticonColony/Assets/Code/Tile.cs]: Calculation of production is already handled within the Roboticon class. Function should be removed.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mart500) [inline RoboticonColony/Assets/Code/Tile.cs]: Combine these 2 into a function

public int GetResources(ItemType resourceType) {
     return resources[resourceType];
}

**Review by Mart500 — CHANGES_REQUESTED** [review]
Issues with redundant functions. If not fixed before merge, wrong addressing of the class could be used else where.

File: RoboticonColony/Assets/Code/Tile.cs
Comment (Mikeywalsh) [reply RoboticonColony/Assets/Code/Tile.cs]: I think we need @RobinStephenson to clear this up a bit. In the UML diagram from assessment one, both Roboticon and Tile have a Produce method in them!

**Review by Mikeywalsh — APPROVED** [review]

**Review by Mikeywalsh — CHANGES_REQUESTED** [review]
Looks good! I think we just need @RobinStephenson to clear up the Produce Method. In the UML diagram it is used in both Tile and Roboticon, but surely it should only be in one class. At the moment, both @Mart500 and @tc975 have their own implementations of it.

**Review by Mikeywalsh — APPROVED** [review]

**Comment by Mikeywalsh** [comment]
Looks good, just waiting on @Mart500 to approve now!

**Comment by Mart500** [comment]
@Mikeywalsh @RobinStephenson @tc975 I just finished cleaning the class up a little. I think it should be ready for merging now.

**Comment by RobinStephenson** [comment]
@Mart500 you need to set yourself as approving the changes for anyone to be able to merge

**Review by Mart500 — APPROVED** [review]

**Comment by Mart500** [comment]
@RobinStephenson Fixed

**Comment by RobinStephenson** [comment]
okay someone just needs to resolve the merge conflicts first

**Pull request merged**
